src/sentry/models/releasefile.py

```diff
@@ -177,45 +194,144 @@ def extract(self) -> TemporaryDirectory:
+class _Manifest:
+    def __init__(self, data: dict, fresh=False):
+        self._data = data
+        self.changed = fresh
+
+    @property
+    def data(self):
+        """Meant to be read-only"""
+        return self._data
+
+    def get(self, filename: str):
+        return self._data.get("files", {}).get(filename, None)
+
+    def update_files(self, files: dict):
+        if files:
+            self._data.setdefault("files", {}).update(files)
+            self.changed = True
+
+    def delete(self, filename: str):
+        self._data.get("files", {}).pop(filename, None)
+        self.changed = True
+
+
+class _ManifestGuard:
+    def __init__(self, release: Release, dist: Optional[Distribution]):
+        self._release = release
+        self._dist = dist
+
+    def readable_data(self) -> Optional[dict]:
+        """Simple read, no synchronization necessary"""
+        file_ = self._get_file(lock=False)
+        if file_ is not None:
+            with file_.getfile() as fp:
+                return json.load(fp)
+
+    @contextmanager
+    def writable_data(self, create: bool):
+        """Context manager for editable release manifest"""
+        with transaction.atomic():
+            if create:
+                file_, created = self._get_or_create_file()
+            else:
+                file_ = self._get_file(lock=True)
+                created = False
+
+            if file_ is None:
+                manifest = None
+            else:
+                if created:
+                    manifest = _Manifest({}, fresh=True)
+                else:
+                    data = json.load(file_.getfile())
+                    manifest = _Manifest(data)
+
+            yield manifest  # editable reference to manifest
+
+            if manifest is not None and manifest.changed:
+                file_.putfile(BytesIO(json.dumps(manifest.data).encode()))
+
+    def _get_or_create_file(self) -> Tuple[File, bool]:
+        qs = ReleaseFile.objects.select_related("file").select_for_update()
+        release_file, created = qs.get_or_create(
+            organization_id=self._release.organization_id,
+            release=self._release,
+            dist=self._dist,
+            name=MANIFEST_FILENAME,
+            file=File.objects.get_or_create(
+                name=MANIFEST_FILENAME,
+                type=MANIFEST_TYPE,
+            )[0],
+        )
+
+        return release_file.file, created
+
+    def _get_file(self, lock: bool) -> Optional[File]:
+        qs = ReleaseFile.objects.select_related("file")
+        if lock:
+            qs = qs.select_for_update()
+        try:
+            release_file = qs.get(
+                organization_id=self._release.organization_id,
+                release=self._release,
+                dist=self._dist,
+                name=MANIFEST_FILENAME,
+            )
+        except ReleaseFile.DoesNotExist:
+            return None
+        else:
+            return release_file.file
+
+
+class ReleaseManifest:
```

Comment:
_Disclaimer_: Assuming "ReleaseIndex" rather than manifest to distinguish between the Sentry-internal concept and what the CLI is sending.

This class is actually a controller class. However, all of its uses are single function invocations. If that is the intended way, let's rather expose functions like: 
- `read_release_index(release, dist)`
- `update_release_index(release, dist, archive)`
- `delete_release_index(release, dist)`